manual/src/refman/expr.etex

```diff
@@ -1292,58 +1268,85 @@ otherwise the default is nonvariant. This is also the case for
 constrained arguments in type definitions.
 
 
-\subsection{ss:expr-other}{Other}
+\subsection{ss:expr-local-items}{Local structure items}
+(Introduced in OCaml 5.5)
 
-\subsubsection*{sss:expr-assertion}{Assertion checking}
+It is possible to introduce any structure item (see~\ref{ss:mexpr-structures})
+defined by @local-definition@ in a local scope: the expression
+\begin{center}
+@"let" local-definition "in" expr@
+\end{center}
+introduces the structure item @local-definition@ during the evaluation of
+@expr@. This construction is equivalent to:
+\begin{center}
+@"let" "module" M = "struct" local-definition "let" body "=" expr "end" "in" M.body@
+\end{center}
+without the overhead of creating the actual module @M@ nor the binding @body@.
 
+The rest of this section contains remarks and examples specific to particular
+cases of this construction.
 
-\ikwd{assert\@\texttt{assert}}
+\subsubsection{sss:expr-let-type}{Local types}
+\ikwd{let\@\texttt{type}}
 
-OCaml supports the @"assert"@ construct to check debugging assertions.
-The expression @"assert" expr@ evaluates the expression @expr@ and
-returns @"()"@ if @expr@ evaluates to @"true"@.  If it evaluates to
-@"false"@ the exception
-"Assert_failure" is raised with the source file name and the
-location of @expr@ as arguments.  Assertion
-checking can be turned off with the "-noassert" compiler option.  In
-this case, @expr@ is not evaluated at all.
+The form @"let" typedef { "and" typedef } "in" expr@ introduces a type (or a
+collection of mutually recursive types) in the scope of an expression:
 
 \begin{caml_example}{toplevel}
-let f a b c =
-  assert (a <= b && b <= c);
-  (b -. a) /. (c -. b);;
+let f x y =
+  let type t = {x: float; y: float} in
+  let norm p = p.x *. p.x +. p.y *. p.y in
+  norm {x; y}
+;;
 \end{caml_example}
 
-As a special case, @"assert false"@ is reduced to
-@'raise' '('@"Assert_failure ..."@')'@, which gives it a polymorphic
-type.  This means that it can be used in place of any expression (for
-example as a branch of any pattern-matching).  It also means that
-the @"assert false"@ ``assertions'' cannot be turned off by the
-"-noassert" option.
-%
-\index{Assertfailure\@\verb`Assert_failure`}
+The compiler checks that values created from this type do not escape this scope:
 
 \begin{caml_example}{toplevel}
-let min_known_nonempty = function
-  | [] -> assert false
-  | l -> List.hd (List.sort compare l);;
+let f x y =
+  let type t = {x: float; y: float} in
+  let twice p = {x = p.x +. p.x; y = p.y +. p.y} in
+  twice {x; y}
+[@@expect error];;
 \end{caml_example}
 
-\subsubsection*{sss:expr-lazy}{Lazy expressions}
-\ikwd{lazy\@\texttt{lazy}}
+Extensible types can also be introduced in this way. The example below exhibits
```

Comment:
Terminology issue: this is demonstrating the creation of an extensible type constructor, not extensible type.
The creation of an extensible type would be `let type t = .. in blabla`.